concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs

```diff
@@ -430,11 +430,21 @@ getWeakPointer weakPtr ptrHash name = do
             lf <- use lastFinalized
             if ptrHash == getHash lf then
               return lf
-            else do
-              nb <- readBlock ptrHash
-              case nb of
-                Just sb -> constructBlock sb
-                Nothing -> logErrorAndThrowTS ("Couldn't find " ++ name ++ " block in disk")
+            else
+              -- Weak pointers are used for parent pointers. A block that is alive should always have
+              -- a parent that is also alive, which means either actually in memory `BlockAlive` or already
+              -- finalized. If we fail to dereference the weak pointer we should thus be able to directly look
+              -- up the block from the block table.
+              use (blockTable . at' ptrHash) >>=
+                 \case Just (BlockAlive bp) -> return bp
```

Comment:
Does this case occur sufficiently often that it is more efficient to check the block table first than to call `readBlock` first?

Reply:
I do not have the numbers unfortunately and when logging the cases it does vary how often it happens based on how it is run so I'm not really sure which option is best.

I did it this way since I assumed that lookup in the hashmap will still be  dominated by a lookup in the LMDB database.